### ISSUE TITLE
Multi-GPU Inference

### ISSUE THREAD

**Issue by derow0208** [huggingface/accelerate#3341]
I trained a classification model using 4 GPUs using acclerate. This is my code for saving the model
```python
if val_accuracy > best_val_accuracy:
    best_val_accuracy = val_accuracy
    accelerator.wait_for_everyone()
    unwrapped_model = accelerator.unwrap_model(model)
    accelerator.save(unwrapped_model.state_dict(), best_model_path) #only model params
    log.info(f"Save best model at epoch: {epoch + 1}!")
```
I now want to test the performance of the model on 2 GPUs，I tried two different ways of loading the model，the first way is：
```python
model.load_state_dict(torch.load(best_model_path)) #load parameters first
# DDP
test_data, model= accelerator.prepare(
    test_data, model
)
# Initialize variables for saving best model
accuracy = 0.0
total_num = 0
correct = 0
# train
model.eval()
with torch.no_grad():
    for i, (img, label) in tqdm(enumerate(test_data), total=len(test_data), disable=not accelerator.is_local_main_process):
        output = model(img)
        
        _, predicted = torch.max(output.data, 1)
        predicted = accelerator.gather_for_metrics(predicted)
        label = accelerator.gather_for_metrics(label)
        
        total_num += label.size(0)
        correct += (predicted == label).sum().item()
            
    accuracy = 100 * correct / total_num
    accelerator.print(f"Testing accuracy: {accuracy:.2f}%")
    accelerator.print(f"Total_num: {total_num}")
    log.info(f"Testing accuracy: {accuracy:.2f}%")
```
the second way is:
```python 
# DDP
  test_data, model= accelerator.prepare(
      test_data, model
  )
  unwrapped_model = accelerator.unwrap_model(model)
  unwrapped_model.load_state_dict(torch.load(best_model_path))
  
  # Initialize variables for saving best model
  accuracy = 0.0
  total_num = 0
  correct = 0
  # train
  model.eval()
  with torch.no_grad():
      for i, (img, label) in tqdm(enumerate(test_data), total=len(test_data), disable=not accelerator.is_local_main_process):
          output = model(img)
          
          _, predicted = torch.max(output.data, 1)
          predicted = accelerator.gather_for_metrics(predicted)
          label = accelerator.gather_for_metrics(label)
          
          total_num += label.size(0)
          correct += (predicted == label).sum().item()
              
      accuracy = 100 * correct / total_num
      accelerator.print(f"Testing accuracy: {accuracy:.2f}%")
      accelerator.print(f"Total_num: {total_num}")
      log.info(f"Testing accuracy: {accuracy:.2f}%")

```
The difference between the two is that the first method loads the parameters first and then uses _accelerate_ to assign the model, while the second method uses _accelerate_ to assign and then uses _unwrapped_model_ to load the parameters. However, the test results of the two ways are completely different. I tested it on the training set. The accuracy of the first way is similar to the accuracy during training, but the result of the second way is much lower than the accuracy during training.

Which one is the correct way to load model? Or are there problems with both? Is there any correct code example for multi-GPU inference?

**Comment by muellerzr** [comment]
The former is the proper way. We have more examples here: https://github.com/huggingface/accelerate/blob/main/examples/inference/distributed/phi2.py

`split_between_processes` is synonymous with preparing a dataloader. 

See also https://huggingface.co/docs/accelerate/usage_guides/distributed_inference

**Comment by derow0208** [comment]
Why is the memory usage of each GPU different during inference? (During training, the memory usage of each GPU is the same)

**Comment by muellerzr** [comment]
They should be quite similar. To what degree? And using the `accelerator.prepare` method I presume? (Without using `device_map`/etc?)

**Comment by derow0208** [comment]
> They should be quite similar. To what degree? And using the `accelerator.prepare` method I presume? (Without using `device_map`/etc?)

Yes, the differences are small and only one GPU will be different from the rest (when tested on 4 GPUs using the code above)，thank you for solving my question.
